pkg/console/controllers/clidownloads/controller.go

```diff
@@ -0,0 +1,285 @@
+)
+
+const (
+	controllerWorkQueueKey = "clidownloads-sync-work-queue-key"
+	controllerName         = "ConsoleCLIDownloadsSyncController"
+
+	ocCLIDownloadsCustomResourceName  = "oc-cli-downloads"
+	odoCLIDownloadsCustomResourceName = "odo-cli-downloads"
+)
+
+type CLIDownloadsSyncController struct {
+	// clients
+	operatorConfigClient      operatorclientv1.ConsoleInterface
+	operatorClient            v1helpers.OperatorClient
+	consoleCliDownloadsClient consoleclientv1.ConsoleCLIDownloadInterface
+	routeClient               routeclientv1.RoutesGetter
+	// listers
+	cachesToSync []cache.InformerSynced
+	queue        workqueue.RateLimitingInterface
+	recorder     events.Recorder
+}
+
+func NewCLIDownloadsSyncController(
+	operatorConfigClient operatorclientv1.OperatorV1Interface,
+	operatorClient v1helpers.OperatorClient,
+	cliDownloadsInterface consoleclientv1.ConsoleCLIDownloadInterface,
+	routev1Client routeclientv1.RoutesGetter,
+	consoleCLIDownloadsInformers consoleinformersv1.ConsoleCLIDownloadInformer,
+	routesInformers routesinformersv1.RouteInformer,
+	recorder events.Recorder,
+) *CLIDownloadsSyncController {
+
+	ctrl := &CLIDownloadsSyncController{
+		// configs
+		operatorConfigClient: operatorConfigClient.Consoles(),
+		// clients
+		consoleCliDownloadsClient: cliDownloadsInterface,
+		routeClient:               routev1Client,
+		// recorder
+		recorder: recorder,
+		queue:    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ConsoleCliDownloadsSyncer"),
+	}
+
+	operatorClient.Informer().AddEventHandler(ctrl.newEventHandler())
+	consoleCLIDownloadsInformers.Informer().AddEventHandler(ctrl.newEventHandler())
+	routesInformers.Informer().AddEventHandler(ctrl.newEventHandler())
+
+	ctrl.cachesToSync = append(ctrl.cachesToSync,
+		operatorClient.Informer().HasSynced,
+		consoleCLIDownloadsInformers.Informer().HasSynced,
+		routesInformers.Informer().HasSynced,
+	)
+
+	return ctrl
+}
+
+func (c *CLIDownloadsSyncController) sync() error {
+	operatorConfig, err := c.operatorConfigClient.Get(api.ConfigResourceName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	updatedOperatorConfig := operatorConfig.DeepCopy()
+
+	switch updatedOperatorConfig.Spec.ManagementState {
+	case operatorsv1.Managed:
+		klog.V(4).Infoln("console is in an managed state: syncing ConsoleCliDownloads custom resources")
+	case operatorsv1.Unmanaged:
+		klog.V(4).Infoln("console is in an unmanaged state: skipping ConsoleCliDownloads custom resources sync")
+		return nil
+	case operatorsv1.Removed:
+		klog.V(4).Infoln("console is in an removed state: deleting ConsoleCliDownloads custom resources")
+		return c.removeCLIDownloads()
+	default:
+		return fmt.Errorf("console is in an unknown state: %v", updatedOperatorConfig.Spec.ManagementState)
+	}
+
+	consoleRoute, err := c.routeClient.Routes(api.TargetNamespace).Get(api.OpenShiftConsoleDownloadsRouteName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
```

Comment:
Pondering if we want 2 different `types`:

```golang
status.HandleDegraded(client, "OCDownloadsSync", ....)
status.HandleDegraded(client, "ODODownloadsSync",....)
```
Since we `return err` immediately after, we can't really stomp over the `oc` sync error with an `odo` sync error. But, in general, I think we don't want to reuse `type` to avoid the potential for stomping on an err (say if we refactor down the road and the shape here changes).

Reply:
Makes sense to me !